velox/exec/tests/PartitionAndSerialize.cpp

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/tests/PartitionAndSerialize.h"
+#include "velox/exec/HashPartitionFunction.h"
+#include "velox/row/UnsafeRowDynamicSerializer.h"
+
+namespace facebook::velox::exec {
+
+void PartitionAndSerializeNode::addDetails(std::stringstream& stream) const {
+  stream << "(";
+  for (auto i = 0; i < keys_.size(); ++i) {
+    const auto& expr = keys_[i];
+    if (i > 0) {
+      stream << ", ";
+    }
+    if (auto field =
+            std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(expr)) {
+      stream << field->name();
+    } else if (
+        auto constant =
+            std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr)) {
+      stream << constant->toString();
+    } else {
+      stream << expr->toString();
+    }
+  }
+  stream << ") " << numPartitions_;
+}
+
+namespace {
+class PartitionAndSerializeOperator : public Operator {
+ public:
+  PartitionAndSerializeOperator(
+      int32_t operatorId,
+      DriverCtx* FOLLY_NONNULL ctx,
+      const std::shared_ptr<const PartitionAndSerializeNode>& planNode)
+      : Operator(
+            ctx,
+            planNode->outputType(),
+            operatorId,
+            planNode->id(),
+            "PartitionAndSerialize") {
+    auto inputType = planNode->sources()[0]->outputType();
+    auto keyChannels = toChannels(inputType, planNode->keys());
+
+    partitionFunction_ = std::make_unique<HashPartitionFunction>(
+        planNode->numPartitions(), inputType, keyChannels);
+  }
+
+  bool needsInput() const override {
+    return !input_;
+  }
+
+  void addInput(RowVectorPtr input) override {
+    input_ = std::move(input);
+  }
+
+  RowVectorPtr getOutput() override {
+    if (!input_) {
+      return nullptr;
+    }
+
+    auto numInput = input_->size();
+
+    // TODO Reuse output vector.
+    auto output = std::dynamic_pointer_cast<RowVector>(
+        BaseVector::create(outputType_, numInput, pool()));
+
+    computePartitions(*output->childAt(0)->asFlatVector<int32_t>());
+
+    serializeRows(*output->childAt(1)->asFlatVector<StringView>());
+
+    input_.reset();
+
+    return output;
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return noMoreInput_;
+  }
+
+ private:
+  void computePartitions(FlatVector<int32_t>& partitionsVector) {
+    auto numInput = input_->size();
+
+    partitions_.resize(numInput);
+    partitionFunction_->partition(*input_, partitions_);
+
+    // TODO Avoid copy.
+    partitionsVector.resize(numInput);
+    auto rawPartitions = partitionsVector.mutableRawValues();
+    std::memcpy(rawPartitions, partitions_.data(), sizeof(int32_t) * numInput);
```

Comment:
Is there another way to write partitions into the raw vectors and avoid the vector copy? I think we can have another version of partition function to do that.

Reply:
Yes, we can have an API that takes a pre-allocated buffer as uint32_t*

```
partitionFunction_->partition(*input_, rawPartitions);
```